internal/docker/docker.go

```diff
@@ -0,0 +1,252 @@
+
+	err := download.FileFromURL(colonyTemplateURL, colonyTemplateYaml)
+	if err != nil {
+		return fmt.Errorf("error downloading file: %w", err)
+	} else {
+		log.Info("downloaded colony.yaml successfully:", colonyTemplateYaml)
+	}
+
+	err = hydrateTemplate(pwd, ColonyTokens{
+		LoadBalancerIP:        loadBalancerIP,
+		LoadBalancerInterface: loadBalancerInterface,
+	})
+
+	if err != nil {
+		return fmt.Errorf("error hydrating template: %w", err)
+	}
+
+	defer c.cli.Close()
+
+	// check for an existing colony-k3s container
+	k3sColonyContainer, err := getColonyK3sContainer(ctx, c)
+	if k3sColonyContainer != nil {
+		return fmt.Errorf("%q container already exists. please remove before continuing or run `colony destroy`", constants.ColonyK3sContainerName)
+	}
+	if err != nil && err != ErrK3sContainerNotFound {
+		return fmt.Errorf("docker error: %w", err)
+	}
+
+	// Pull the rancher/k3s image if it’s not already available
+	imageName := "rancher/k3s:v1.30.2-k3s1"
+	reader, err := c.cli.ImagePull(ctx, imageName, image.PullOptions{})
+	if err != nil {
+		return fmt.Errorf("error pulling image %q: %w", imageName, err)
+	}
+	log.Infof("Pulled image %q successfully", imageName)
+
+	defer reader.Close()
+	// c.cli.ImagePull is asynchronous.
+	// The reader needs to be read completely for the pull operation to complete.
+	// If stdout is not required, consider using io.Discard instead of os.Stdout.
+	io.Copy(os.Stdout, reader)
+
+	env := []string{
+		fmt.Sprintf("K3S_KUBECONFIG_OUTPUT=%s", constants.KubeconfigDockerPath),
+		"K3S_KUBECONFIG_MODE=666",
+	}
+
+	mounts := []mount.Mount{
+		{
+			Type:   mount.TypeBind,
+			Source: pwd,
+			Target: "/output",
+		},
+		{
+			Type:   mount.TypeVolume,
+			Source: "k3s-server",
+			Target: "/var/lib/rancher/k3s",
+		},
+		{
+			Type:   mount.TypeBind,
+			Source: filepath.Join(pwd, constants.ColonyYamlPath),
+			Target: fmt.Sprintf("/var/lib/rancher/k3s/server/manifests/%s", constants.ColonyYamlPath),
+		},
+		{
+			Type:   mount.TypeTmpfs,
+			Target: "/run",
+		},
+		{
+			Type:   mount.TypeTmpfs,
+			Target: "/var/run",
+		},
+	}
+
+	resp, err := c.cli.ContainerCreate(ctx, &containerTypes.Config{
+		Image: imageName,
+		Env:   env,
+		Cmd: []string{
+			"server",
+			"--disable=traefik,servicelb",
+			"--tls-san=colony",
+			"--node-label=colony.konstruct.io/node-type=colony",
+		},
+	}, &containerTypes.HostConfig{
+		Privileged:  true,
+		NetworkMode: "host",
+		Mounts:      mounts,
+	}, nil, nil, constants.ColonyK3sContainerName)
+
+	if err != nil {
+		log.Error("Error creating container: %w", err)
+	}
+
+	log.Infof("created container with ID %q", resp.ID)
+
+	if err := c.cli.ContainerStart(ctx, resp.ID, containerTypes.StartOptions{}); err != nil {
+		panic(err)
+	}
+
+	waitInterval := 2 * time.Second
+	timeout := 15 * time.Second
+
+	log.Infof("Checking for file %s every %d...\n", filepath.Join(pwd, constants.KubeconfigHostPath), waitInterval)
```

Comment:
Log lines should never end on a new line. That's because the logger can be replaced for a structured logger, a JSON logger, a remote logger or everything in between.

```suggestion
	log.Infof("Checking for file %s every %d...", filepath.Join(pwd, constants.KubeconfigHostPath), waitInterval)
```